app/models/upload.rb

```diff
@@ -1,37 +1,80 @@
 # frozen_string_literal: true
 class Upload < ActiveRecord::Base
-  attr_accessor :skip_lines, :upload_file
+  attr_accessor :skip_lines, :upload_file, :missing_headers, :extra_headers
 
   belongs_to :user, inverse_of: :versions
 
   validates_associated :user
   validates :user_id, presence: true
 
   validates :csv, presence: true
+
   validate :csv_type_check?
 
+  after_initialize :initialize_warnings, unless: :persisted?
   after_initialize :derive_dependent_columns, unless: :persisted?
 
+  def derive_dependent_columns
+    self.csv = upload_file.try(:original_filename)
+  end
+
   def ok?
     ok
   end
 
+  def all_warnings
+    missing_headers.full_messages + extra_headers.full_messages
+  end
+
+  def self.last_uploads
+    max_query = Upload.select('csv_type, MAX(updated_at) as max_updated_at').where(ok: true).group(:csv_type).to_sql
+
+    joins("INNER JOIN (#{max_query}) max_uploads ON uploads.csv_type = max_uploads.csv_type")
+      .where('uploads.updated_at = max_uploads.max_updated_at')
+      .order(:csv_type)
+  end
+
   def csv_type_check?
     return true if InstitutionBuilder::TABLES.map(&:name).push('Institution').include?(csv_type)
 
-    errors.add(:csv_type, "#{csv_type} is not a valid CSV data source")
+    if csv_type.present?
+      errors.add(:csv_type, "#{csv_type} is not a valid CSV data source")
+    else
+      errors.add(:csv_type, 'cannot be blank.')
+    end
+
     false
   end
 
-  def derive_dependent_columns
-    self.csv = upload_file.try(:original_filename)
+  def check_for_headers
+    return unless upload_file && csv_type && skip_lines
+
+    missing_headers.clear
```

Comment:
@patrickvinograd I think we share differing opinions on this one. I am the one advocating for this change vs Marc's initial PR in which does precisely what you describe. 

The notices or warnings are specific to the upload instance, and I think it's best to keep that logic at a level where it can be easily unit tested vs at the controller level. I'm also a strong believer in the "thin" controller paradigm. I prefer cruft to be in the model or service level objects. 

That Flashable Mixin was a mess, and for me it's a welcome improvement to get rid of it.